### ISSUE TITLE
macos x86_64, win64 on py3.10 are failing lib-python test_ssl.py

### ISSUE THREAD

**Issue by mattip** [pypy/pypy#4883]
These may be important for a functioning python? Passing `test_ssl.py` is one of the acceptance tests for the conda-forge package, so it is currently blocked.

On macos:
```
======================================================================
ERROR: test_get_server_certificate (test.test_ssl.SimpleBackgroundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/test/test_ssl.py", line 2219, in test_get_server_certificate
    _test_get_server_certificate(self, *self.server_addr, cert=SIGNING_CA)
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/test/test_ssl.py", line 2436, in _test_get_server_certificate
    pem = ssl.get_server_certificate((host, port), ca_certs=cert)
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/ssl.py", line 1578, in get_server_certificate
    with create_connection(addr, timeout=timeout) as sock:
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/socket.py", line 845, in create_connection
    raise err
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/socket.py", line 833, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

======================================================================
ERROR: test_get_server_certificate_sni (test.test_ssl.SimpleBackgroundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/test/test_ssl.py", line 2234, in test_get_server_certificate_sni
    pem = ssl.get_server_certificate((host, port), ca_certs=SIGNING_CA)
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/ssl.py", line 1578, in get_server_certificate
    with create_connection(addr, timeout=timeout) as sock:
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/socket.py", line 845, in create_connection
    raise err
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/socket.py", line 833, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

======================================================================
ERROR: test_msg_callback_deadlock_bpo43577 (test.test_ssl.TestSSLDebug)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/test/test_ssl.py", line 4943, in test_msg_callback_deadlock_bpo43577
    s.connect((HOST, server.port))
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/ssl.py", line 1430, in connect
    self._real_connect(addr, False)
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/ssl.py", line 1421, in _real_connect
    self.do_handshake()
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib-python/3/ssl.py", line 1397, in do_handshake
    self._sslobj.do_handshake()
  File "/Users/matti/build-worker-x86_64/pypy-c-jit-macos-x86-64/build/lib_pypy/_cffi_ssl/_stdssl/__init__.py", line 535, in do_handshake
    raise pyssl_error(self, ret)
ConnectionResetError: [Errno 54] Connection reset by peer

----------------------------------------------------------------------
Ran 179 tests in 45.423s

FAILED (errors=3, skipped=23)
test_ssl failed (3 errors) in 46.0 sec
```

On windows:
```
======================================================================
ERROR: test_get_server_certificate (__main__.SimpleBackgroundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "D:\bld\pypy3.10_1706562248406\_test_env\Lib\test\test_ssl.py", line 2234, in test_get_server_certificate_sni
    pem = ssl.get_server_certificate((host, port), ca_certs=SIGNING_CA)
  File "D:\bld\pypy3.10_1706562248406\_test_env\Lib\ssl.py", line 1578, in get_server_certificate
    with create_connection(addr, timeout=timeout) as sock:
  File "D:\bld\pypy3.10_1706562248406\_test_env\Lib\socket.py", line 845, in create_connection
    raise err
  File "D:\bld\pypy3.10_1706562248406\_test_env\Lib\socket.py", line 833, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 10061] No connection could be made because the target machine actively refused it

======================================================================
FAIL: test_load_default_certs_env_windows (__main__.ContextTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\bld\pypy3.10_1706562248406\_test_env\Lib\test\test_ssl.py", line 1691, in test_load_default_certs_env_windows
    self.assertEqual(ctx.cert_store_stats(), stats)
AssertionError: {'x509': 417, 'x509_ca': 409, 'crl': 0} != {'x509': 418, 'x509_ca': 409, 'crl': 0}
- {'crl': 0, 'x509': 417, 'x509_ca': 409}
?                      ^

+ {'crl': 0, 'x509': 418, 'x509_ca': 409}
?                      ^


----------------------------------------------------------------------
Ran 179 tests in 41.568s

FAILED (failures=1, errors=2, skipped=22)

```

**Comment by mattip** [comment]
The call to `test_get_server_certificate` is failing locally on linux x86_64 as well. After the connection is established, the client socket dies and the server prints "Connection reset by peer"

**Comment by mattip** [comment]
The server that prints the message is running in a separate thread. I suspect some race condition is causing the client's `socket.close()` to happen in the middle of a transaction causing the server to crash, but I haven't been able to pinpoint it yet.

**Comment by mattip** [comment]
Fixed in eb1aa836f1eb21ded16a5110b7adf941c77326a6, but now I think tests that were passing before now fail. Let's see what the buildbot brings

**Comment by mattip** [comment]
Yup, I fixed a bit and broke a bit. Now there are two failing `test_bio_*` tests. Here is one on x86_64:
```
$ pypy/goal/pypy3.10-c lib-python/3/test/test_ssl.py -k test_bio_handshake -vv
test_ssl: testing with 'OpenSSL 3.0.2 15 Mar 2022' (3, 0, 0, 2, 0)
          under 'Linux-5.15.0-91-generic-x86_64-with-glibc2.35'
          HAS_SNI = True
          OP_ALL = 0x80000050
          OP_NO_TLSv1_1 = 0x10000000
test_bio_handshake (__main__.SimpleBackgroundTests) ...  server:  new connection from ('127.0.0.1', 43892)
Needed 2 calls to complete do_handshake().
 server: connection cipher is now ('TLS_AES_256_GCM_SHA384', 'TLSv1.3', 256)
Test server failure:
Traceback (most recent call last):
   File "/home/matti/oss/pypy/lib-python/3/test/test_ssl.py", line 2560, in run
    msg = self.read()
   File "/home/matti/oss/pypy/lib-python/3/test/test_ssl.py", line 2537, in read
    return self.sslconn.read()
   File "/home/matti/oss/pypy/lib-python/3/ssl.py", line 1165, in read
    return self._sslobj.read(len)
   File "/home/matti/oss/pypy/lib_pypy/_cffi_ssl/_stdssl/__init__.py", line 660, in read
    return self._read_no_buf(length)
   File "/home/matti/oss/pypy/lib_pypy/_cffi_ssl/_stdssl/__init__.py", line 716, in _read_no_buf
    raise pyssl_error(self, retval)
 _cffi_ssl._stdssl.error.SSLZeroReturnError: TLS/SSL connection has been closed (EOF)
ERROR

======================================================================
ERROR: test_bio_handshake (__main__.SimpleBackgroundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/matti/oss/pypy/lib-python/3/test/test_ssl.py", line 2366, in test_bio_handshake
    self.ssl_io_loop(sock, incoming, outgoing, sslobj.unwrap)
  File "/home/matti/oss/pypy/lib-python/3/test/test_ssl.py", line 2313, in ssl_io_loop
    ret = func(*args)
  File "/home/matti/oss/pypy/lib-python/3/ssl.py", line 979, in unwrap
    return self._sslobj.shutdown()
  File "/home/matti/oss/pypy/lib_pypy/_cffi_ssl/_stdssl/__init__.py", line 939, in shutdown
    raise pyssl_error(self, err)
_cffi_ssl._stdssl.error.SSLError: [SSL: APPLICATION_DATA_AFTER_CLOSE_NOTIFY] application data after close notify

----------------------------------------------------------------------
Ran 1 test in 0.050s

FAILED (errors=1)
```

and, for comparison, here is CPython 3.10

```
$ python3.10 lib-python/3/test/test_ssl.py -k test_bio_handshake -vv
test_ssl: testing with 'OpenSSL 3.0.2 15 Mar 2022' (3, 0, 0, 2, 0)
          under 'Linux-5.15.0-91-generic-x86_64-with-glibc2.35'
          HAS_SNI = True
          OP_ALL = 0x80000050
          OP_NO_TLSv1_1 = 0x10000000
test_bio_handshake (__main__.SimpleBackgroundTests) ...  server:  new connection from ('127.0.0.1', 46142)
Needed 2 calls to complete do_handshake().
 server: connection cipher is now ('TLS_AES_256_GCM_SHA384', 'TLSv1.3', 256)
Needed 4 calls to complete unwrap().
 connection timeout TimeoutError('timed out')
ok

----------------------------------------------------------------------
Ran 1 test in 1.014s

OK
```

**Comment by mattip** [comment]
Should be fixed in 4fc60db. Closing for now
